In [183]:
# pip install spacy==2.3.5
# python -m spacy download en
# pip install neuralcoref --no-binary neuralcoref (got binary error on github so this install)
# !pip install spacy-experimental

# neuralcoref==4.0 - not used anymore 
# spacy==2.3.5 - old version to support neuralcoref


# Python==3.8 (3.7.10 backup plan)

#!pip install https://github.com/explosion/spacy-experimental/releases/download/v0.6.1/en_coreference_web_trf-3.4.0a2-py3-none-any.whl
!pip install transformers
!pip install fastcoref


huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


In [184]:

# Load your usual SpaCy model (one of SpaCy English models)
# import spacy
# nlp = spacy.load('en')

# Add neural coref to SpaCy's pipe
# import neuralcoref
# neuralcoref.add_to_pipe(nlp)
# 
# # You're done. You can now use NeuralCoref as you usually manipulate a SpaCy document annotations.
# doc = nlp(u'My sister has a dog. She loves him.')
# 
# doc._.has_coref
# doc._.coref_clusters

In [185]:
# # Access the resolved coreferences
# for token in doc:
#     print(token.text, token._.in_coref, token._.coref_clusters)
# 
# # Iterate over resolved coreference clusters
# for cluster in doc._.coref_clusters:
#     print(cluster.mentions)

In [7]:
import trafilatura
downloaded = trafilatura.fetch_url('https://github.blog/2019-03-29-leader-spotlight-erin-spiceland/')
trafilatura.extract(downloaded)

"Release Radar · Hacktoberfest 2023 Edition\nAs Hacktoberfest comes to a close, you still have a day or so left to complete and submit your pull requests. If you're struggling to think of projects to contribute…\nWe’re spending Women’s History Month with women leaders who are making history every day in the tech community. Read more about Erin Spiceland: Software Engineer at SpaceX.\nEvery March we recognize the women who have shaped history—and now, we’re taking a look forward. From driving software development in large companies to maintaining thriving open source communities, we’re spending Women’s History Month with women leaders who are making history every day in the tech community. Erin Spiceland is a Software Engineer for SpaceX. Born and raised in rural south Georgia, she is a Choctaw and Chickasaw mother of two now living in downtown Los Angeles. Erin didn’t finish college—she’s a predominantly self-taught software engineer. In her spare time, she makes handmade Native Americ

In [8]:
from trafilatura import feeds
# mylist = feeds.find_feed_urls('https://www.dailymail.co.uk/news/uk-politics/index.rss')
mylist = feeds.find_feed_urls('https://www.dailymail.co.uk/news/rishi-sunak/index.rss')
print(mylist)

['https://www.dailymail.co.uk/?ito=1490&ns_campaign=1490&ns_mchannel=rss', 'https://www.dailymail.co.uk/debate/article-12637749/QUENTIN-LETTS-minutes-silence-Commons-Hamas-atrocity.html?ito=1490&ns_campaign=1490&ns_mchannel=rss', 'https://www.dailymail.co.uk/debate/article-12642127/ALEX-BRUMMER-utterly-dispiriting-Tories-chosen-hike-taxes-cynical-way-damaging-Britain.html?ito=1490&ns_campaign=1490&ns_mchannel=rss', 'https://www.dailymail.co.uk/debate/article-12646767/QUENTIN-LETTS-PMQs-pious-contest-sound-sepulchral.html?ito=1490&ns_campaign=1490&ns_mchannel=rss', 'https://www.dailymail.co.uk/debate/article-12651425/Not-generation-world-felt-dangerously-febrile-freedoms-enemies-Russia-China-Iran-exploit-spiralling-crisis-qualified-answer-MARK-ALMOND-doesnt-pull-punches.html?ito=1490&ns_campaign=1490&ns_mchannel=rss', 'https://www.dailymail.co.uk/debate/article-12655049/andrew-neil-rishi-sunak-conservatives-labour-election-tory-election.html?ito=1490&ns_campaign=1490&ns_mchannel=rss', '

https://adrien.barbaresi.eu/blog/how-to-download-parallel-politeness-rules-python.html
Provides a method to check the robot txt of a site and check if the url being extracted meets the politeness policy.

When extracting from bing check for each domain and store in data structure with allow/disallow. If the same base url appears again then no need to recheck robot.txt

In [9]:
import urllib.robotparser

# define a website to look for rules
base_url = "https://www.theguardian.com"

# load the necessary components, fetch and parse the file
rules = urllib.robotparser.RobotFileParser()
rules.set_url(base_url + "/robots.txt")
rules.read()

print(rules)

# determine if a page can be fetched by all crawlers
rules.can_fetch("*", "https://www.theguardian.com/import urllib.robotparser")
# returns True or False

User-agent: Mediapartners-Google
Allow: 

User-agent: NewsNow
Disallow: /

User-agent: GPTBot
Disallow: /

User-agent: CCBot
Disallow: /

User-agent: TurnitinBot
Disallow: /

User-agent: PetalBot
Disallow: /

User-agent: MoodleBot
Disallow: /

User-agent: *
Disallow: /sendarticle/
Disallow: /Users/
Disallow: /users/
Disallow: /%2A/print%24
Disallow: /email/
Disallow: /contactus/
Disallow: /share/
Disallow: /websearch
Disallow: /%2A%3Fcommentpage%3D
Disallow: /whsmiths/
Disallow: /external/overture/
Disallow: /discussion/report-abuse/%2A
Disallow: /discussion/report-abuse-ajax/%2A
Disallow: /discussion/comment-permalink/%2A
Disallow: /discussion/report-abuse/%2A
Disallow: /discussion/user-report-abuse/%2A
Disallow: /discussion/handlers/%2A
Disallow: /discussion/your-profile
Disallow: /discussion/your-comments
Disallow: /discussion/edit-profile
Disallow: /discussion/search/comments
Disallow: /discussion/%2A
Disallow: /search
Disallow: /music/artist/%2A
Disallow: /music/album/%2A
Disallow

True

In [10]:
downloaded = trafilatura.fetch_url(mylist[6])
trafilatura.extract(downloaded, favour_precision=True, include_comments=False, include_images=False, include_tables=False)

"We Tories are in a torpor not seen since Henry VI's catatonic state, writes SIR JACOB REES-MOGG\nThe Government is in a torpor not seen since Henry the Sixth was in a catatonic stupor, unable to speak or govern for 18 months in the 15th century.\nNothing is being done to inspire voters to turn out, and to assume that they will do so next year to stop Sir Keir Starmer is wishful thinking.\nSir Keir has been busy putting on Tory clothes, learning from the electorally-successful Sir Tony Blair.\nIn such a way, he appears much less threatening than either Corbyn or Miliband.\nYet. Henry the Sixth did come out of his stupor and it is not yet impossible for the Tories to do the same.\nIn recent weeks, some of the Government's rhetoric has improved a little.\nSir Keir has been busy putting on Tory clothes, learning from the electorally-successful Sir Tony Blair. In such a way, he appears much less threatening than previous Labour leaders Jeremy Corbyn or Ed Miliband. However, Henry VI did co

In [11]:
downloaded = trafilatura.fetch_url('https://www.dailymail.co.uk/news/article-12350195/Think-twice-imposing-Ulez-tax-hard-working-Brits-Rishi-Sunak-blasts-Sadiq-Khan-hated-low-emission-zone-expansion-plan.html?ns_mchannel=rss&ns_campaign=1490&ito=1490')
article_text = trafilatura.extract(downloaded)

In [177]:
file_path = "ULEZ-daily-mail-research-article.txt"

# Write the text to the file
with open(file_path, 'w', encoding='utf-8') as file:
    file.write(article_text)

print(f"Article text saved to {file_path}")

Article text saved to ULEZ-daily-mail-research-article.txt


In [178]:
file_path = "ULEZ-daily-mail-research-article.txt"

# Read the text from the file
with open(file_path, 'r', encoding='utf-8') as file:
    article_text = file.read()

# Now, article_text contains the text you read from the file
print(f"Article text read from {file_path}:\n{article_text}")    

Article text read from ULEZ-daily-mail-research-article.txt:
'Think twice before imposing Ulez tax on hard-working Brits': Rishi Sunak blasts Sadiq Khan for his hated low emission zone expansion plan
- Keir Starmer told to 'get off the fence' over Sadiq Khan's Ulez by senior Tory MPs
- Rishi Sunak led MPs asking Starmer to get London mayor to postpone the levy
- READ MORE: Labour civil war over ULEZ intensifies after Khan's High Court win
Keir Starmer faced fresh calls last night for him to try to delay the expansion of London's clean air zone.
Senior Tories said he should 'get off the fence' and force the capital's mayor Sadiq Khan to protect motorists set to pay the Ulez levy.
The High Court yesterday backed the extension of Ulez into outer London – a move that could lead to similar schemes elsewhere in Britain.
Sir Keir has refused to say whether the zone should expand, claiming it is not a 'simple political decision'. He remained silent after yesterday's court ruling.
Rishi Sunak l

NLTK Experimentation - Unused Currently.

In [12]:

# # Download the NLTK Punkt tokenizer data if not already downloaded
# nltk.download('punkt')
# 
# from nltk.tokenize import sent_tokenize
# import nltk
# 
# # Download the NLTK Punkt tokenizer data if not already downloaded
# nltk.download('punkt')
# 
# from nltk.tokenize import sent_tokenize
# 
# # Split the text into sentences
# sentences = sent_tokenize(article_text)
# 
# # Print the sentences
# for sentence in sentences:
#     print("sentence:" + sentence)


In [128]:
import spacy
nlp = spacy.load("en_coreference_web_trf")
doc = nlp("The cats were startled by the dog as it growled at them.") 
print(doc.spans)

/Users/ameek/anaconda3/envs/ProjectProposal/lib/python3.8/site-packages/spacy/util.py:910: UserWarning: [W095] Model 'en_coreference_web_trf' (3.4.0a2) was trained with spaCy v3.3.0 and may not be 100% compatible with the current version (3.6.1). If you see errors or degraded performance, download a newer compatible model or retrain your custom model with the current spaCy version. For more details and available updates, run: python -m spacy validate
  warnings.warn(warn_msg)


{'coref_clusters_1': [the dog, it], 'coref_clusters_2': [The cats, them]}


In [129]:
import spacy
doc = nlp("This is a sentence.")
coref = nlp.add_pipe("experimental_coref")
# This usually happens under the hood
processed = coref(doc)

KeyError: "Parameter 'E' for model 'hashembed' has not been allocated yet."

In [130]:
import torch

In [131]:
# Check Metal Performance Shaders (MPS) backend for GPU training acceleration.
if torch.backends.mps.is_available():
    mps_device = torch.device("mps")
    x = torch.ones(1, device=mps_device)
    print (x)
else:
    print ("MPS device not found.")

tensor([1.], device='mps:0')


In [132]:
from fastcoref import FCoref

model = FCoref(device='mps')

preds = model.predict(
   texts=['We are so happy to see you using our coref package. This package is very fast!']
)

preds[0].get_clusters(as_strings=False)


print(preds[0].get_clusters())


preds[0].get_logit(
   span_i=(33, 50), span_j=(52, 64)
)


11/01/2023 22:43:16 - INFO - 	 missing_keys: []
11/01/2023 22:43:16 - INFO - 	 unexpected_keys: []
11/01/2023 22:43:16 - INFO - 	 mismatched_keys: []
11/01/2023 22:43:16 - INFO - 	 error_msgs: []
11/01/2023 22:43:16 - INFO - 	 Model Parameters: 90.5M, Transformer: 82.1M, Coref head: 8.4M
11/01/2023 22:43:16 - INFO - 	 Tokenize 1 inputs...


Map:   0%|          | 0/1 [00:00<?, ? examples/s]

11/01/2023 22:43:16 - INFO - 	 ***** Running Inference on 1 texts *****


Inference:   0%|          | 0/1 [00:00<?, ?it/s]

[['We', 'our'], ['our coref package', 'This package']]


18.852913

Playing around with extracting coreference clusters using fastcoref directly
"Processing can be applied to a collection of texts of any length in a batched and parallel fashion"
https://pypi.org/project/fastcoref/

In [133]:
from fastcoref import FCoref

model = FCoref(device='mps')

preds = model.predict(
   texts=[article_text]
)

clustersText = preds[0].get_clusters()
clustersPositions = preds[0].get_clusters(as_strings=False)

print(clustersText)

combined_clusters = [(text, positions, len(text)) for text, positions in zip(clustersText, clustersPositions)]

sorted_combined_clusters = sorted(combined_clusters, key=lambda x: x[2], reverse=True)

for i, (clusterText, clusterPos, cluster_text_count) in enumerate(sorted_combined_clusters, 1):
    print(f"Cluster {i} : {clusterText}")
    print(f"Cluster Text Count: {cluster_text_count}")
    print("Positions:", clusterPos)


11/01/2023 22:43:19 - INFO - 	 missing_keys: []
11/01/2023 22:43:19 - INFO - 	 unexpected_keys: []
11/01/2023 22:43:19 - INFO - 	 mismatched_keys: []
11/01/2023 22:43:19 - INFO - 	 error_msgs: []
11/01/2023 22:43:19 - INFO - 	 Model Parameters: 90.5M, Transformer: 82.1M, Coref head: 8.4M
11/01/2023 22:43:19 - INFO - 	 Tokenize 1 inputs...


Map:   0%|          | 0/1 [00:00<?, ? examples/s]

11/01/2023 22:43:20 - INFO - 	 ***** Running Inference on 1 texts *****


Inference:   0%|          | 0/1 [00:00<?, ?it/s]

[['Sadiq Khan', 'his', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'I', 'I', 'Mr Khan', 'Sadiq', 'He', 'he', 'his', 'he', 'your mayor', 'the Labour mayor', 'his', "Mr Khan's", "the mayor's", 'his', 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"], ['Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak (pictured)'], ['Keir Starmer', 'Starmer', 'Keir Starmer', 'him', 'he', 'Sir Keir', 'He', 'the Labour leader', 'Keir Starmer (left)', 'him', 'the Labour leader', "Sir Keir Starmer's", 'Starmer', 'Sir Keir', 'the Labour leader', 'Keir', 'he', 'Sir Keir', 'Sir Keir'], ["Sadiq Khan's Ulez", 'ULEZ', 'ULEZ'], ['London', "London's", "the capital's", 'London', "London's", 'London', 'London', "London's", 'London', "London's", 'London'], ['the levy', 'the Ulez levy', 'the £12.50 daily levy for older, more polluting vehicles', 'it', 'the £12.50 dai

In [134]:
from fastcoref import LingMessCoref

model = LingMessCoref(device='cpu')

preds = model.predict(
   texts=[article_text]
)

print(preds[0].get_clusters()) 

11/01/2023 22:43:30 - INFO - 	 missing_keys: []
11/01/2023 22:43:30 - INFO - 	 unexpected_keys: []
11/01/2023 22:43:30 - INFO - 	 mismatched_keys: []
11/01/2023 22:43:30 - INFO - 	 error_msgs: []
11/01/2023 22:43:30 - INFO - 	 Model Parameters: 590.0M, Transformer: 434.6M, Coref head: 155.4M
11/01/2023 22:43:30 - INFO - 	 Tokenize 1 inputs...


Map:   0%|          | 0/1 [00:00<?, ? examples/s]

11/01/2023 22:43:30 - INFO - 	 ***** Running Inference on 1 texts *****


Inference:   0%|          | 0/1 [00:00<?, ?it/s]

[['Sadiq Khan', 'his', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'I', 'I', 'Sir Keir', 'the Labour leader', 'Mr Khan', 'Keir', 'he', 'Sadiq', 'He', 'he', 'his', 'he', 'Sir Keir', 'your mayor', 'Sir Keir', 'the Labour mayor', 'his', "Mr Khan's", "the mayor's", 'his', 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"], ['Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak (pictured)'], ['Keir Starmer', 'Starmer', 'Keir Starmer', 'him', 'he', 'Sir Keir', 'He', 'the Labour leader', 'Keir Starmer (left)', 'him', 'the Labour leader', "Sir Keir Starmer's", 'Starmer'], ['Ulez', 'ULEZ', 'Ulez', 'Ulez', 'Ulez', 'ULEZ', 'Ulez', 'the Ulez', 'Ulez', 'Ulez', 'the policy', 'Ulez', 'Ulez', 'Ulez', 'Ulez', 'Ulez', 'Ulez', 'Ulez', 'the scheme', 'Ulez'], ['London', "London's", "the capital's", 'London', 'London', "London's", 'London', 'London', "London

In [24]:
# TODO: look into spacy pipelines, spans, etc

In [25]:
from fastcoref import spacy_component
# Cell not required anymore

import spacy

text = article_text 

nlp = spacy.load("en_core_web_sm")
nlp.add_pipe("fastcoref")

doc = nlp(text)
doc._.coref_clusters

10/31/2023 16:31:48 - INFO - 	 missing_keys: []
10/31/2023 16:31:48 - INFO - 	 unexpected_keys: []
10/31/2023 16:31:48 - INFO - 	 mismatched_keys: []
10/31/2023 16:31:48 - INFO - 	 error_msgs: []
10/31/2023 16:31:48 - INFO - 	 Model Parameters: 90.5M, Transformer: 82.1M, Coref head: 8.4M
10/31/2023 16:31:49 - INFO - 	 Tokenize 1 inputs...


Map:   0%|          | 0/1 [00:00<?, ? examples/s]

10/31/2023 16:31:49 - INFO - 	 ***** Running Inference on 1 texts *****


Inference:   0%|          | 0/1 [00:00<?, ?it/s]

[[(81, 91),
  (96, 99),
  (187, 199),
  (268, 280),
  (360, 366),
  (549, 579),
  (1006, 1013),
  (1230, 1255),
  (1360, 1367),
  (1479, 1503),
  (1735, 1745),
  (2099, 2108),
  (2503, 2510),
  (2669, 2670),
  (2688, 2689),
  (3393, 3400),
  (3741, 3746),
  (3771, 3773),
  (3796, 3798),
  (3866, 3869),
  (3936, 3938),
  (4136, 4146),
  (5326, 5342),
  (5352, 5355),
  (5915, 5924),
  (5999, 6010),
  (6154, 6157),
  (6342, 6349),
  (7103, 7110),
  (7216, 7225),
  (7468, 7479)],
 [(62, 73), (226, 237), (926, 937), (1269, 1291)],
 [(141, 153),
  (253, 260),
  (382, 394),
  (428, 431),
  (509, 511),
  (765, 773),
  (875, 877),
  (980, 997),
  (1140, 1159),
  (1193, 1196),
  (1334, 1351),
  (1479, 1497),
  (1688, 1695),
  (3075, 3083),
  (3291, 3308),
  (3613, 3617),
  (3643, 3645),
  (4086, 4094),
  (5279, 5287)],
 [(187, 204), (337, 341), (1639, 1643)],
 [(268, 274),
  (465, 473),
  (549, 562),
  (2342, 2348),
  (2891, 2899),
  (3541, 3547),
  (3763, 3769),
  (4309, 4317),
  (5790, 5796),



But to my current knowledge: The spaCy library itself does not natively support processing an array of text documents in a single call in parallel.
So consider applying fastCoref directly then using spaCy or alternative to extract sentences

https://medium.com/quantrium-tech/top-3-packages-for-named-entity-recognition-e9e14f6f0a2a
    
And so decided to play around with flair

In [64]:
# !pip install flair

In [ ]:
# Decided to explore more than just org and person detections - comment out this cell:
# import spacy
# from flair.data import Sentence
# from flair.models import SequenceTagger
# 
# NER = spacy.load("en_core_web_sm")
# 
# article_text = [article_text][0]
# 
# article = NER(article_text)
# org_entities = {'ORG': []}
# person_entities = {'PERSON': []}
# 
# for word in article.ents:
#     label = word.label_
#     if label in org_entities:
#         org_entities[label].append([[word.text, word.start_char, word.end_char], label])
#     if label in person_entities:
#         person_entities[label].append([[word.text, word.start_char, word.end_char], label])
# 
# # Print the ORG entities
# print("ORG Entities:")
# for label, entities in org_entities.items():
#     for entity in entities:
#         entity_text, start_pos, end_pos = entity[0]
#         label = entity[1]
#         print(f"Entity: {entity_text}, Position: [{start_pos}, {end_pos}], Label: {label}")
# 
# # Print the PERSON entities
# print("PERSON Entities:")
# for label, entities in person_entities.items():
#     for entity in entities:
#         entity_text, start_pos, end_pos = entity[0]
#         label = entity[1]
#         print(f"Entity: {entity_text}, Position: [{start_pos}, {end_pos}], Label: {label}")

In [135]:
from flair.data import Sentence
from flair.models import SequenceTagger

# List of entities
entity_list = [
    'Sadiq Khan', 'his', "Sadiq Khan's", 'London mayor', "Khan's",
    "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's",
    'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan',
    'I', 'I', 'Mr Khan', 'Sadiq', 'He', 'he', 'his', 'he', 'your mayor',
    'the Labour mayor', 'his', "Mr Khan's", "the mayor's", 'his', 'Mr Khan',
    'Mr Khan', 'the mayor', "the mayor's"
]

# Create a sentence with the entities
sentence = Sentence(" ".join(entity_list))

# Load Flair's NER model (e.g., English NER model)
tagger = SequenceTagger.load("ner")

# Predict NER tags for the sentence
tagger.predict(sentence)

# Extract NER tags and entities
ner_tags = sentence.to_dict(tag_type="ner")["entities"]

# Print the NER tags and entities
for entity in ner_tags:
    print(f"Entity: {entity['text']}, Label: {entity['type']}")



2023-11-01 22:43:39,477 SequenceTagger predicts: Dictionary with 20 tags: <unk>, O, S-ORG, S-MISC, B-PER, E-PER, S-LOC, B-ORG, E-ORG, I-PER, S-PER, B-MISC, I-MISC, E-MISC, I-ORG, B-LOC, E-LOC, I-LOC, <START>, <STOP>


KeyError: 'type'

In [136]:
MENTION_REQ_PER = 0.2
import spacy
import math
from functools import reduce

NER = spacy.load("en_core_web_sm")

article_text = [article_text][0]

article = NER(article_text)

num_sentences = len(list(article.sents))

threshold = math.floor(num_sentences * MENTION_REQ_PER)

# Print the results
print(f"Number of Sentences: {num_sentences}")
print(f"Coreference Threshold: {threshold}")


def merge_positions(entities, word):
    entity_key = word.text + word.label_
    if entity_key in entities:
        entities[entity_key][1].append([word.start_char, word.end_char])
    else:
        entities[entity_key] = [word.text, [[word.start_char, word.end_char]], word.label_]
    return entities

org_entities = reduce(merge_positions, filter(lambda word: word.label_ == 'ORG', article.ents), {})
person_entities = reduce(merge_positions, filter(lambda word: word.label_ == 'PERSON', article.ents), {})

# Print the merged ORG entities
print("Merged ORG Entities:")
for entity_key, entity_info in org_entities.items():
    entity_text = entity_info[0]
    positions = entity_info[1]
    label = entity_info[2]
    print(f"Entity: {entity_text}, Positions: {positions}, Label: {label}")

# Print the merged PERSON entities
print("Merged PERSON Entities:")
for entity_key, entity_info in person_entities.items():
    entity_text = entity_info[0]
    positions = entity_info[1]
    label = entity_info[2]
    print(f"Entity: {entity_text}, Positions: {positions}, Label: {label}")


Number of Sentences: 59
Coreference Threshold: 11
Merged ORG Entities:
Entity: Ulez, Positions: [[29, 33], [200, 204], [616, 620], [676, 680], [1573, 1577], [2627, 2631], [2967, 2971], [3342, 3346], [4182, 4186], [5356, 5360], [5650, 5654], [5925, 5929], [7350, 7354]], Label: ORG
Entity: Starmer, Positions: [[253, 260]], Label: ORG
Entity: ULEZ, Positions: [[337, 341], [1639, 1643]], Label: ORG
Entity: High Court, Positions: [[367, 377], [1802, 1812]], Label: ORG
Entity: The High Court, Positions: [[627, 641], [5739, 5753]], Label: ORG
Entity: Labour, Positions: [[984, 990], [1338, 1344], [3169, 3175], [3295, 3301], [5330, 5336], [6554, 6560]], Label: ORG
Entity: RAC, Positions: [[2212, 2215], [7202, 7205]], Label: ORG
Entity: Driver, Positions: [[2413, 2419]], Label: ORG
Entity: Vehicle Licensing Agency, Positions: [[2424, 2448]], Label: ORG
Entity: the High Court, Positions: [[2805, 2819]], Label: ORG
Entity: Times Radio, Positions: [[3599, 3610]], Label: ORG
Entity: Energy, Position

In [179]:
import spacy
import math
from functools import reduce

NER = spacy.load("en_core_web_sm")

article_text = [article_text][0]

article = NER(article_text)

for ent in article.ents:
    print(f"Entity: {ent.text}, Start: {ent.start_char}, End: {ent.end_char}, Label: {ent.label_}")

Entity: Ulez, Start: 29, End: 33, Label: ORG
Entity: Brits, Start: 54, End: 59, Label: NORP
Entity: Rishi, Start: 62, End: 67, Label: PERSON
Entity: Sadiq Khan, Start: 81, End: 91, Label: PERSON
Entity: Keir Starmer, Start: 141, End: 153, Label: PERSON
Entity: Sadiq Khan's, Start: 187, End: 199, Label: FAC
Entity: Ulez, Start: 200, End: 204, Label: ORG
Entity: Tory, Start: 215, End: 219, Label: PERSON
Entity: Starmer, Start: 253, End: 260, Label: ORG
Entity: London, Start: 268, End: 274, Label: GPE
Entity: ULEZ, Start: 337, End: 341, Label: ORG
Entity: Khan, Start: 360, End: 364, Label: PERSON
Entity: High Court, Start: 367, End: 377, Label: ORG
Entity: Keir Starmer, Start: 382, End: 394, Label: PERSON
Entity: last night, Start: 413, End: 423, Label: TIME
Entity: London, Start: 465, End: 471, Label: GPE
Entity: Sadiq Khan, Start: 569, End: 579, Label: PERSON
Entity: Ulez, Start: 616, End: 620, Label: ORG
Entity: The High Court, Start: 627, End: 641, Label: ORG
Entity: yesterday, Start:

Possible NER outcomes: CARDINAL, DATE, EVENT, FAC, GPE, LANGUAGE, LAW, LOC, MONEY, NORP, ORDINAL, ORG, PERCENT, PERSON, PRODUCT, QUANTITY, TIME, WORK_OF_ART
https://huggingface.co/spacy/en_core_web_sm

In [138]:
def merge_positions(entities, word):
    entity_key = (word.text + word.label_).lower()
    if entity_key in entities:
        entities[entity_key][1].append([word.start_char, word.end_char])
    else:
        entities[entity_key] = [word.text, [[word.start_char, word.end_char]], word.label_]
    return entities

org_entities = reduce(merge_positions, filter(lambda word: word.label_ == 'ORG', article.ents), {})
person_entities = reduce(merge_positions, filter(lambda word: word.label_ == 'PERSON', article.ents), {})


print("Merged ORG Entities:")
for entity_key, entity_info in org_entities.items():
    entity_text = entity_info[0]
    positions = entity_info[1]
    label = entity_info[2]
    print(f"Entity: {entity_text}, Positions: {positions}, Label: {label}")


print("Merged PERSON Entities:")
for entity_key, entity_info in person_entities.items():
    entity_text = entity_info[0]
    positions = entity_info[1]
    label = entity_info[2]
    print(f"Entity: {entity_text}, Positions: {positions}, Label: {label}")
    


Merged ORG Entities:
Entity: Ulez, Positions: [[29, 33], [200, 204], [337, 341], [616, 620], [676, 680], [1573, 1577], [1639, 1643], [2627, 2631], [2967, 2971], [3342, 3346], [4182, 4186], [5356, 5360], [5650, 5654], [5925, 5929], [7350, 7354]], Label: ORG
Entity: Starmer, Positions: [[253, 260]], Label: ORG
Entity: High Court, Positions: [[367, 377], [1802, 1812]], Label: ORG
Entity: The High Court, Positions: [[627, 641], [2805, 2819], [5739, 5753]], Label: ORG
Entity: Labour, Positions: [[984, 990], [1338, 1344], [3169, 3175], [3295, 3301], [5330, 5336], [6554, 6560]], Label: ORG
Entity: RAC, Positions: [[2212, 2215], [7202, 7205]], Label: ORG
Entity: Driver, Positions: [[2413, 2419]], Label: ORG
Entity: Vehicle Licensing Agency, Positions: [[2424, 2448]], Label: ORG
Entity: Times Radio, Positions: [[3599, 3610]], Label: ORG
Entity: Energy, Positions: [[4010, 4016]], Label: ORG
Entity: Londoners, Positions: [[5404, 5413]], Label: ORG
Entity: Surrey County Council, Positions: [[5859,

NER could be confused and additional mentions could be used to determine "correct" category

In [180]:
import spacy
from functools import reduce

NER = spacy.load("en_core_web_sm")

article_text = [article_text][0]

article = NER(article_text)

# num_sentences = len(list(article.sents))
# 
# 
# # Store sentence positions for use with coreference mentions and news sentiment later
# sentence_positions = [(sent.start_char, sent.end_char) for sent in article.sents]

# # Print sentences + start and end pos.
# for i, (start, end) in enumerate(sentence_positions):
#     sentence_text = article[start:end].text
#     print(f"Sentence {i+1}: {sentence_text}")
#     print(f"Start Position: {start}, End Position: {end}")
#     print()


entity_types = ["CARDINAL", "DATE", "EVENT", "FAC", "GPE", "LANGUAGE", "LAW", "LOC", "MONEY", "NORP", "ORDINAL", "ORG", "PERCENT", "PERSON", "PRODUCT", "QUANTITY", "TIME", "WORK_OF_ART"]

entity_type_to_entities = {
    entity_type: [
        [
            entity_text,
            positions,
            label,
            len(positions)  # Count of positions
        ] for entity_text, positions, label in reduce(
            merge_positions,
            filter(lambda word: word.label_ == entity_type, article.ents),
            {}
        ).values()
    ] for entity_type in entity_types
}

# print(entity_type_to_entities)

for entity_type, entities in entity_type_to_entities.items():
    print(f"Entity Type: {entity_type}")
    for entity in entities:
        entity_text, positions, label, num_positions = entity
        print(f"Entity: {entity_text}")
        print(f"Positions: {positions}")
        print(f"Label: {label}")
        print(f"Number of Positions: {num_positions}")
        print()
        
people_entities = {entity_type: entity_info for entity_type, entity_info in entity_type_to_entities.items() if entity_type == 'PERSON'}

        

for entity in entity_type_to_entities[entity_type]:
    entity_text, positions, label, num_positions = entity
    if entity_type not in people_entities:
        people_entities[entity_type] = []
    people_entities[entity_type].append({
        'Entity': entity_text,
        'Positions': positions,
        'Label': label,
        'Number of Positions': num_positions
    })

Entity Type: CARDINAL
Entity: nearly 700,000
Positions: [[2131, 2145]]
Label: CARDINAL
Number of Positions: 1

Entity: More than 1.5million
Positions: [[2299, 2319]]
Label: CARDINAL
Number of Positions: 1

Entity: nine
Positions: [[2520, 2524]]
Label: CARDINAL
Number of Positions: 1

Entity: ten
Positions: [[2532, 2535]]
Label: CARDINAL
Number of Positions: 1

Entity: five
Positions: [[6309, 6313]]
Label: CARDINAL
Number of Positions: 1

Entity: Seven
Positions: [[6392, 6397]]
Label: CARDINAL
Number of Positions: 1

Entity Type: DATE
Entity: yesterday
Positions: [[642, 651], [900, 909], [2483, 2492]]
Label: DATE
Number of Positions: 3

Entity: daily
Positions: [[1041, 1046], [1395, 1400], [1768, 1773], [2181, 2186], [4268, 4273], [4628, 4633]]
Label: DATE
Number of Positions: 6

Entity: August 29
Positions: [[1129, 1138]]
Label: DATE
Number of Positions: 1

Entity: 29 August
Positions: [[1786, 1795]]
Label: DATE
Number of Positions: 1

Entity: weekly
Positions: [[1946, 1952]]
Label: DA

In [181]:
for entity_type, entities in people_entities.items():
    print(f"Entity Type Contained in Dict: {entity_type}")
    for entity in entities:
        entity_text = entity[0]
        positions = entity[1]
        label = entity[2]
        num_positions = entity[3]
        print(f"Entity: {entity_text}")
        print(f"Positions: {positions}")
        print(f"Label: {label}")
        print(f"Number of Positions: {num_positions}")
        print()


Entity Type Contained in Dict: PERSON
Entity: Rishi
Positions: [[62, 67], [1269, 1274]]
Label: PERSON
Number of Positions: 2

Entity: Sadiq Khan
Positions: [[81, 91], [569, 579], [1243, 1253], [1735, 1745]]
Label: PERSON
Number of Positions: 4

Entity: Keir Starmer
Positions: [[141, 153], [382, 394], [1140, 1152], [1483, 1495]]
Label: PERSON
Number of Positions: 4

Entity: Tory
Positions: [[215, 219]]
Label: PERSON
Number of Positions: 1

Entity: Khan
Positions: [[360, 364], [1009, 1013], [1363, 1367], [2102, 2106], [2506, 2510], [3396, 3400], [5918, 5922], [6345, 6349], [7106, 7110]]
Label: PERSON
Number of Positions: 9

Entity: Keir
Positions: [[769, 773], [3079, 3083], [3613, 3617], [4090, 4094], [5283, 5287]]
Label: PERSON
Number of Positions: 5

Entity: Rishi Sunak
Positions: [[926, 937]]
Label: PERSON
Number of Positions: 1

Entity: Starmer
Positions: [[1688, 1695]]
Label: PERSON
Number of Positions: 1

Entity: Boris Johnson's
Positions: [[3034, 3049]]
Label: PERSON
Number of Pos

In [182]:
print(sorted_combined_clusters[1][2])
print(people_entities)

19
{'PERSON': [['Rishi', [[62, 67], [1269, 1274]], 'PERSON', 2], ['Sadiq Khan', [[81, 91], [569, 579], [1243, 1253], [1735, 1745]], 'PERSON', 4], ['Keir Starmer', [[141, 153], [382, 394], [1140, 1152], [1483, 1495]], 'PERSON', 4], ['Tory', [[215, 219]], 'PERSON', 1], ['Khan', [[360, 364], [1009, 1013], [1363, 1367], [2102, 2106], [2506, 2510], [3396, 3400], [5918, 5922], [6345, 6349], [7106, 7110]], 'PERSON', 9], ['Keir', [[769, 773], [3079, 3083], [3613, 3617], [4090, 4094], [5283, 5287]], 'PERSON', 5], ['Rishi Sunak', [[926, 937]], 'PERSON', 1], ['Starmer', [[1688, 1695]], 'PERSON', 1], ["Boris Johnson's", [[3034, 3049]], 'PERSON', 1], ['Rachel Reeves', [[3466, 3479], [3696, 3709]], 'PERSON', 2], ['Wes Streeting', [[3579, 3592]], 'PERSON', 1], ['Grant Shapps', [[4027, 4039]], 'PERSON', 1], ['Ulez', [[4200, 4204]], 'PERSON', 1], ['Trader', [[4811, 4817]], 'PERSON', 1], ['Ian Plummer', [[4996, 5007]], 'PERSON', 1], ['Lucy Frazer', [[5259, 5270]], 'PERSON', 1], ['Mark Harper', [[5436, 5

In [183]:
#  sorted_combined_clusters
MENTION_REQ_PER = 0.07 # Keep as this for ULEZ research article findings later on.

threshold = math.floor(num_sentences * MENTION_REQ_PER)

# Print the results
print(f"Number of Sentences: {num_sentences}")
print("Coreference threshold: {:.2f} ({}% detected cluster mentions versus sentences)".format(threshold, int(MENTION_REQ_PER * 100)))
print("Cluster Entities before filtration: ", len(sorted_combined_clusters))

filtered_clusters = [(i, cluster) for i, cluster in enumerate(sorted_combined_clusters) if cluster[2] >= threshold]

print("Cluster Entities after filtration: ", len(filtered_clusters))

print(filtered_clusters)

Number of Sentences: 59
Coreference threshold: 4.00 (7% detected cluster mentions versus sentences)
Cluster Entities before filtration:  38
Cluster Entities after filtration:  14
[(0, (['Sadiq Khan', 'his', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'I', 'I', 'Mr Khan', 'Sadiq', 'He', 'he', 'his', 'he', 'your mayor', 'the Labour mayor', 'his', "Mr Khan's", "the mayor's", 'his', 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"], [(81, 91), (96, 99), (187, 199), (268, 280), (360, 366), (549, 579), (1006, 1013), (1230, 1255), (1360, 1367), (1479, 1503), (1735, 1745), (2099, 2108), (2503, 2510), (2669, 2670), (2688, 2689), (3393, 3400), (3741, 3746), (3771, 3773), (3796, 3798), (3866, 3869), (3936, 3938), (4136, 4146), (5326, 5342), (5352, 5355), (5915, 5924), (5999, 6010), (6154, 6157), (6342, 6349), (7103, 7110), (7216, 7225), (7468, 7479)]

In [92]:
# entity_to_cluster_mapping = {}
# 
# for entity_type, entities in people_entities.items():
#     for entity in entities:
#         entity_name = entity[0]
#         for cluster in filtered_clusters:
#             cluster_text = cluster[0]
#             for coref_entry in cluster_text:
#                 if entity_name in coref_entry:
#                     if entity_name not in entity_to_cluster_mapping:
#                         entity_to_cluster_mapping[entity_name] = []
# 
#                     entity_to_cluster_mapping[entity_name].append((cluster_text, cluster[1]))
#                     break
# print(entity_to_cluster_mapping) 
# Replaced by beneath code 

Entity to cluster mapping initially was is this NER entity containined in any coreference cluster.
This was too simplistic for example:

{'Entity Name': 'Keir Starmer', 'Positions': [[141, 153], [382, 394], [1140, 1152], [1483, 1495]], 'Label': 'PERSON', 'Num Positions': 4, 'Cluster Info': [{'Cluster Text': ['Sadiq Khan', 'his', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'I', 'I', 'Mr Khan', 'Sadiq', 'He', 'he', 'his', 'he', 'your mayor', 'the Labour mayor', 'his', "Mr Khan's", "the mayor's", 'his', 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"]. Yes 'Keir' is in the coreference mention 'Sir Keir Starmer's mayor' but on balance wouldn't you agree these coreferences are not referring to him.


Solutions I have thought about: 
1. Removing pro-nouns (his, her, he, I etc) then doing a % match rate on entity text across the cluster. In the example above we would expect Sadiq / Khan to win out.

2. Removing pro-nouns (his, her, he, I etc) then doing a % match rate on entity position values across the cluster positions.

In [184]:
entity_to_cluster_mapping = []

undesired_words = ["i", "he", "his", "she", "they", "it", "this", "that", "these", "those", "the", "a", "an", "of"]


def cleanse_cluster_text(cluster_text):
    return [word.strip() for word in cluster_text if word.lower().strip() not in undesired_words]

threshold_percentage = 0.15 # If 

for entity_type, entities in people_entities.items():
    for entity in entities:
        entity_name, positions, label, num_positions = entity
        entity_entry = {
            'Entity Name': entity_name,
            'Positions': positions,
            'Label': label,
            'Num Positions': num_positions,
            'Cluster Info': []
        }

        for index, (cluster_text, cluster_positions, _) in filtered_clusters:

            # for coref_entry in cluster_text:
            #     if entity_name in coref_entry:
            #         entity_entry['Cluster Info'] = cluster_entry
            #         break
           
            cleaned_cluster_text = cleanse_cluster_text(cluster_text)
            total_coref_words = " ".join(cleaned_cluster_text)
            entity_count = total_coref_words.count(entity_name)
            percentage = entity_count / len(cleaned_cluster_text)
            if percentage > 0.00:
                print(entity_name)
                print(cleaned_cluster_text)
                print(percentage)
                print('------------------------------------------')
        
            if percentage >= threshold_percentage:
                cluster_entry = {                
                'Cluster Text': cluster_text,
                'Cluster Positions': cluster_positions
            }
                
                entity_entry['Cluster Info'] = cluster_entry
                break

        entity_to_cluster_mapping.append(entity_entry)


Rishi
['Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak (pictured)']
1.0
------------------------------------------
Sadiq Khan
['Sadiq Khan', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'Mr Khan', 'Sadiq', 'your mayor', 'the Labour mayor', "Mr Khan's", "the mayor's", 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"]
0.22727272727272727
------------------------------------------
Keir Starmer
['Sadiq Khan', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'Mr Khan', 'Sadiq', 'your mayor', 'the Labour mayor', "Mr Khan's", "the mayor's", 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"]
0.045454545454545456
------------------------------------------
Keir Starmer
['Keir Starmer', 'Starmer', 'Ke

Evaluation for threshold:

Rishi
['Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak (pictured)']
1.0 Correct
------------------------------------------
Sadiq Khan
['Sadiq Khan', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'Mr Khan', 'Sadiq', 'your mayor', 'the Labour mayor', "Mr Khan's", "the mayor's", 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"]
0.22727272727272727 Correct
------------------------------------------
Sadiq Khan
["the expansion of London's clean air zone", "the expansion of London mayor Sadiq Khan's (right) Ulez", 'expansion', 'the Ulez extension', 'the policy', 'expand', 'the Ulez expansion', 'his Ulez expansion', 'expand', 'expand', "the expansion of London's clean air zone", 'the change', 'Ulez expansion']
0.07692307692307693 Not correct
------------------------------------------
Sadiq Khan
["London's clean air zone", 'Ulez', 'the zone', 'The zone', "London mayor Sadiq Khan's (right) Ulez", 'the expanded Ulez zone', 'the Ulez', 'Ulez', 'Ulez', 'Ulez', 'Ulez', 'Ulez', 'Ulez', 'the scheme', "London's clean air zone", 'Ulez']
0.0625 Not correct
------------------------------------------
Keir Starmer
['Sadiq Khan', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'Mr Khan', 'Sadiq', 'your mayor', 'the Labour mayor', "Mr Khan's", "the mayor's", 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"]
0.045454545454545456 Not correct
------------------------------------------
Keir Starmer
['Keir Starmer', 'Starmer', 'Keir Starmer', 'him', 'Sir Keir', 'the Labour leader', 'Keir Starmer (left)', 'him', 'the Labour leader', "Sir Keir Starmer's", 'Starmer', 'Sir Keir', 'the Labour leader', 'Keir', 'Sir Keir', 'Sir Keir']
0.25 Correct
------------------------------------------
Khan
['Sadiq Khan', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'Mr Khan', 'Sadiq', 'your mayor', 'the Labour mayor', "Mr Khan's", "the mayor's", 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"]
0.6363636363636364 Correct
------------------------------------------
Keir
['Sadiq Khan', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'Mr Khan', 'Sadiq', 'your mayor', 'the Labour mayor', "Mr Khan's", "the mayor's", 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"]
0.045454545454545456 Not correct
------------------------------------------
Keir
['Keir Starmer', 'Starmer', 'Keir Starmer', 'him', 'Sir Keir', 'the Labour leader', 'Keir Starmer (left)', 'him', 'the Labour leader', "Sir Keir Starmer's", 'Starmer', 'Sir Keir', 'the Labour leader', 'Keir', 'Sir Keir', 'Sir Keir']
0.5625 Correct
------------------------------------------
Rishi Sunak
['Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak (pictured)']
1.0 Correct
------------------------------------------
Starmer
['Sadiq Khan', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'Mr Khan', 'Sadiq', 'your mayor', 'the Labour mayor', "Mr Khan's", "the mayor's", 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"]
0.045454545454545456 Not correct
------------------------------------------
Starmer
['Keir Starmer', 'Starmer', 'Keir Starmer', 'him', 'Sir Keir', 'the Labour leader', 'Keir Starmer (left)', 'him', 'the Labour leader', "Sir Keir Starmer's", 'Starmer', 'Sir Keir', 'the Labour leader', 'Keir', 'Sir Keir', 'Sir Keir']
0.375 Correct
------------------------------------------
Rachel Reeves
['Shadow Chancellor Rachel Reeves', 'Rachel Reeves', 'Mr Justice Swift']
0.6666666666666666 Correct
------------------------------------------
Ulez
["the expansion of London's clean air zone", "the expansion of London mayor Sadiq Khan's (right) Ulez", 'expansion', 'the Ulez extension', 'the policy', 'expand', 'the Ulez expansion', 'his Ulez expansion', 'expand', 'expand', "the expansion of London's clean air zone", 'the change', 'Ulez expansion']
0.38461538461538464 Correct
------------------------------------------
Ulez
["London's clean air zone", 'Ulez', 'the zone', 'The zone', "London mayor Sadiq Khan's (right) Ulez", 'the expanded Ulez zone', 'the Ulez', 'Ulez', 'Ulez', 'Ulez', 'Ulez', 'Ulez', 'Ulez', 'the scheme', "London's clean air zone", 'Ulez']
0.6875 Correct
------------------------------------------
On this whole entity name approach with the above data I figure 0.15. Let's see if we can get a higher threshold value somehow... 

Let us try spliting the entity names into part words e.g. Sadiq Khan will be split into Sadiq and Khan for evaluation purposes.

In [186]:
entity_to_cluster_mapping = []

undesired_words = ["i", "he", "his", "she", "they", "it", "this", "that", "these", "those", "the", "a", "an", "of"]


def cleanse_cluster_text(cluster_text):
    return [word.strip() for word in cluster_text if word.lower().strip() not in undesired_words]

threshold_percentage = 0.30

for entity_type, entities in people_entities.items():
    for entity in entities:
        entity_name, positions, label, num_positions = entity
        entity_entry = {
            'Entity Name': entity_name,
            'Positions': positions,
            'Label': label,
            'Num Positions': num_positions,
            'Cluster Info': []
        }

        cluster_id = 0

        for index, (cluster_text, cluster_positions, _) in filtered_clusters:
            
            cluster_id += 1
            
            cleaned_cluster_text = cleanse_cluster_text(cluster_text)
            total_coref_words = " ".join(cleaned_cluster_text)
            
            entity_parts = entity_name.split()
            max_percentage = 0.0

            for entity_part in entity_parts:
                entity_count = total_coref_words.count(entity_part)
                percentage = entity_count / len(cleaned_cluster_text)
            
                if percentage > max_percentage:
                    max_percentage = percentage
                    
                if percentage > 0.00:
                    print("Entity Part: " + entity_part)
                    print(cleaned_cluster_text)
                    print(percentage)
                    print('------------------------------------------')

            if max_percentage > 0.00:
                print("Entity Name: " + entity_name)
                print(cleaned_cluster_text)
                print("Max % Match: " + str(max_percentage))
                if max_percentage > threshold_percentage:
                    print("PREDICTION:  MATCH")
                else:
                    print ("PREDICTION: INVALID")
                print('------------------------------------------')
                   
                    
            cleaned_cluster_text = cleanse_cluster_text(cluster_text)
            total_coref_words = " ".join(cleaned_cluster_text)
            entity_count = total_coref_words.count(entity_name)
            percentage = entity_count / len(cleaned_cluster_text)
            
        
            if percentage >= threshold_percentage:
                cluster_entry = {  
                'Cluster ID' : cluster_id,
                'Cluster Text': cluster_text,
                'Cluster Positions': cluster_positions
            }
                
                entity_entry['Cluster Info'] = cluster_entry
                entity_to_cluster_mapping.append(entity_entry)
                break

        


Entity Part: Rishi
['Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak (pictured)']
1.0
------------------------------------------
Entity Name: Rishi
['Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak (pictured)']
Max % Match: 1.0
PREDICTION:  MATCH
------------------------------------------
Entity Part: Sadiq
['Sadiq Khan', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'Mr Khan', 'Sadiq', 'your mayor', 'the Labour mayor', "Mr Khan's", "the mayor's", 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"]
0.2727272727272727
------------------------------------------
Entity Part: Khan
['Sadiq Khan', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'Mr Khan', 'Sadiq', 'your mayor', 'the Labour

On this new evaluation will double threshold now to p = 0.3 -- could further explore requirement on the start/end position values in text

In [187]:
# print(entity_to_cluster_mapping)

# Bin entities that were not successful in coreference combination stage i.e there cluster info is blank.

for entry in entity_to_cluster_mapping:
    print(entry)
    print()

{'Entity Name': 'Rishi', 'Positions': [[62, 67], [1269, 1274]], 'Label': 'PERSON', 'Num Positions': 2, 'Cluster Info': {'Cluster ID': 11, 'Cluster Text': ['Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak (pictured)'], 'Cluster Positions': [(62, 73), (226, 237), (926, 937), (1269, 1291)]}}

{'Entity Name': 'Khan', 'Positions': [[360, 364], [1009, 1013], [1363, 1367], [2102, 2106], [2506, 2510], [3396, 3400], [5918, 5922], [6345, 6349], [7106, 7110]], 'Label': 'PERSON', 'Num Positions': 9, 'Cluster Info': {'Cluster ID': 1, 'Cluster Text': ['Sadiq Khan', 'his', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'I', 'I', 'Mr Khan', 'Sadiq', 'He', 'he', 'his', 'he', 'your mayor', 'the Labour mayor', 'his', "Mr Khan's", "the mayor's", 'his', 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"], 'Cluster Positions': [(81, 91), (96, 99), (187, 199

We need to consolidate some entries that have been determined as the same cluster ID.

Two types identified so far:
    1. where an entity is the substring of another e.g. Rishi is a substring of Rishi Sunak so should be merged into the 'better' later version with full name.
    2. where an entity first and second name are separate entity lists but one is the first name e.g., Keir and the other is the second name Starmer. I propose trying a combination of 2 of the entities under the same cluster ID and checking if they match any coreference values in cluster text 
    (validating that first / second name combination) so then we can break out create a new combined entity and remove the first / second name instance.


In [188]:
from collections import defaultdict

cluster_dict = defaultdict(list)

for entry in entity_to_cluster_mapping:
    entity_name = entry['Entity Name']
    cluster_id = entry['Cluster Info']['Cluster ID']
    
    cluster_dict[cluster_id].append(entry)


def is_substring(entity1, entity2):
    return entity1 in entity2 or entity2 in entity1

def combine_entities(entities, cluster_text):
    combined_entity = None

    for entity1 in entities:
        for entity2 in entities:
            if entity1 != entity2:
                combined1 = entity1 + ' ' + entity2
                combined2 = entity2 + ' ' + entity1

                if combined1 in cluster_text or combined2 in cluster_text:
                    combined_entity = combined1 if combined1 in cluster_text else combined2
                    break

    return combined_entity

for cluster_id, entries in cluster_dict.items():
    if len(entries) > 1:
        combined_entry = None
        for i, entry1 in enumerate(entries):
            for j, entry2 in enumerate(entries):
                if i < j:
                    entity_1_name = entry1['Entity Name']
                    entity_2_name = entry2['Entity Name']
                    
                    cluster_text = entries[0]['Cluster Info']['Cluster Text']
                    combined_entity = combine_entities([entity_1_name, entity_2_name], cluster_text)
                    
                    
                    if is_substring(entity_1_name, entity_2_name):
                        if len(entity_1_name) > len(entity_2_name):
                            entries.remove(entry2)
                            entry1['Num Positions'] = int(-200) # Coreference cluster will be used anyway -200 to indicate merge via removal.
                            entry1['Positions'] = int(-200) 
                        else:
                            entries.remove(entry1)
                            entry2['Num Positions'] = int(-200) # Coreference cluster will be used anyway -200 to indicate merge via removal.
                            entry2['Positions'] = int(-200) 
                            
                    elif combined_entity in cluster_text:
                        combined_entry = {
                            'Entity Name': combined_entity,
                            'Positions': int(-100), # Coreference cluster will be used anyway -100 to indicate merge via combined entity.
                            'Label': entry1['Label'], 
                            'Num Positions': int(-100),  # Coreference cluster will be used anyway -100 to indicate merge via combined entity.
                            'Cluster Info': entry1['Cluster Info'] 
                        }
                        entries.remove(entry1)
                        entries.remove(entry2)
                        break
        if combined_entry:
            entries.append(combined_entry)                

clustered_entries = [entry for entries in cluster_dict.values() for entry in entries]

for entry in clustered_entries:
    print(entry)
    print()


{'Entity Name': 'Rishi Sunak', 'Positions': -200, 'Label': 'PERSON', 'Num Positions': -200, 'Cluster Info': {'Cluster ID': 11, 'Cluster Text': ['Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak (pictured)'], 'Cluster Positions': [(62, 73), (226, 237), (926, 937), (1269, 1291)]}}

{'Entity Name': 'Khan', 'Positions': [[360, 364], [1009, 1013], [1363, 1367], [2102, 2106], [2506, 2510], [3396, 3400], [5918, 5922], [6345, 6349], [7106, 7110]], 'Label': 'PERSON', 'Num Positions': 9, 'Cluster Info': {'Cluster ID': 1, 'Cluster Text': ['Sadiq Khan', 'his', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'I', 'I', 'Mr Khan', 'Sadiq', 'He', 'he', 'his', 'he', 'your mayor', 'the Labour mayor', 'his', "Mr Khan's", "the mayor's", 'his', 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"], 'Cluster Positions': [(81, 91), (96, 99), (187, 199), (268, 28

With Khan may be worth searching the cluster text for two word mentions that ft. khan to extract Sadiq Khan as a "better" entity name.

In [190]:
# !pip install textblob

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 636.8/636.8 kB 1.2 MB/s eta 0:00:00a 0:00:01


In [363]:
# ***Replaced with TextBlob code below as it had a convenient sentence.start and sentence.end)


# import nltk
# from nltk.tokenize import sent_tokenize
# 
# sentences = sent_tokenize(article_text)
# sentence_boundaries = [(0, len(sentences[0]))] 
# for i in range(1, len(sentences)):
#     start = sentence_boundaries[-1][1] + 1
#     end = start + len(sentences[i])
#     sentence_boundaries.append((start, end))
# 
# 
# # print(sentence_boundaries)
# # Cluster positions (assuming they are character positions within the text)
# cluster_positions = [(20, 35)]
# 
# # Initialise a list to store the selected sentences
# selected_sentences = []
# 
# # Map cluster positions to sentences
# for start, end in cluster_positions:
#     for i, (sentence_start, sentence_end) in enumerate(sentence_boundaries):
#         if sentence_start <= start < sentence_end or sentence_start <= end < sentence_end:
#             selected_sentences.append(sentences[i])
# 
# # Print the selected sentences
# for sentence in selected_sentences:
#     print(sentence)

In [307]:
def insert_intervals(initial_list, new_values):
    def insert_recursive(intervals, values):
        if not values:
            return intervals

        value = values[0]
        result = []
        for interval in intervals:
            if interval[0] <= value <= interval[1]:
                if interval[0] < value:
                    result.append((interval[0], value - 1))
                if value < interval[1]:
                    result.append((value, interval[1]))
            else:
                result.append(interval)
        return insert_recursive(result, values[1:])

    updated_list = insert_recursive(initial_list, new_values)
    return updated_list

In [319]:
def insert_intervals(initial_list, new_values):
    def insert_recursive(intervals, values):
        if not values:
            return intervals

        value = values[0]
        result = []
        for interval in intervals:
            if interval[0] <= value <= interval[1]:
                if interval[0] < value:
                    result.append((interval[0], value))         # To mess around with intervals change value + - offset here.
                if value < interval[1]:
                    result.append((value+1, interval[1]))       # To mess around with intervals change value + - offset here.
            else:
                result.append(interval)
        return insert_recursive(result, values[1:])

    updated_list = insert_recursive(initial_list, new_values)
    return updated_list

In [359]:
from textblob import TextBlob # spaCy was not satisfactory for accurately tokenizing for sentence start / end characters. Trying textblob instead...
import re
from intervaltree import IntervalTree

blob = TextBlob(article_text)

# Get sentences
sentences = blob.sentences

hyphen_sentences = re.split(r'\n-', article_text) # In testing article new line and '-' hyphen use typically means consider a different sentence - SHOULD review other articles.
hyphen_nl_pos = [pos for pos, char in enumerate(article_text) if article_text[pos:pos+2] == '\n-']

extra_split = hyphen_nl_pos

# sentence_bounds = [int(sentence.start) for sentence in sentences] + [int(sentence.end) for sentence in sentences]

sentence_bounds = [(int(sentence.start), int(sentence.end))  for sentence in sentences]

print(sentence_bounds)

extra_split = [138, 223, 301]

updated_list = insert_intervals(sentence_bounds, extra_split)
print(updated_list)

sentence_bounds = updated_list

# Dealing with hyphen-splitting
# start_offset = 0
# for hyphen_sentence in hyphen_sentences:
#     end_offset = start_offset + len(hyphen_sentence)
#     sentence_bounds.append((start_offset, end_offset))
#     start_offset = end_offset
    


# print(sentences)

# for num in extra_split:
#     for i, bound in enumerate(sentence_bounds):
#         if num <= bound:
#             sentence_bounds.insert(i, num)
#             break


print(len(sentence_bounds))






[(0, 489), (490, 626), (627, 764), (765, 874), (875, 925), (926, 1108), (1109, 1139), (1140, 1605), (1606, 1963), (1964, 2098), (2099, 2298), (2299, 2468), (2469, 2768), (2769, 2937), (2938, 3074), (3075, 3220), (3221, 3273), (3274, 3447), (3448, 3548), (3549, 3710), (3711, 3736), (3737, 3770), (3771, 3820), (3821, 3887), (3888, 3931), (3932, 4009), (4010, 4198), (4199, 4396), (4397, 4607), (4608, 4805), (4806, 4981), (4982, 5077), (5078, 5163), (5164, 5240), (5241, 5371), (5372, 5415), (5416, 5621), (5622, 5738), (5739, 5954), (5955, 6166), (6167, 6254), (6255, 6391), (6392, 6553), (6554, 6613), (6614, 6770), (6771, 6956), (6957, 7268), (7269, 7444), (7445, 7604), (7605, 7691), (7692, 7735), (7736, 7856), (7857, 7909)]
[(0, 138), (139, 223), (224, 301), (302, 489), (490, 626), (627, 764), (765, 874), (875, 925), (926, 1108), (1109, 1139), (1140, 1605), (1606, 1963), (1964, 2098), (2099, 2298), (2299, 2468), (2469, 2768), (2769, 2937), (2938, 3074), (3075, 3220), (3221, 3273), (3274, 3

In [321]:
print(sentence_bounds)
print(extra_split)

[(0, 138), (139, 223), (224, 301), (302, 489), (490, 626), (627, 764), (765, 874), (875, 925), (926, 1108), (1109, 1139), (1140, 1605), (1606, 1963), (1964, 2098), (2099, 2298), (2299, 2468), (2469, 2768), (2769, 2937), (2938, 3074), (3075, 3220), (3221, 3273), (3274, 3447), (3448, 3548), (3549, 3710), (3711, 3736), (3737, 3770), (3771, 3820), (3821, 3887), (3888, 3931), (3932, 4009), (4010, 4198), (4199, 4396), (4397, 4607), (4608, 4805), (4806, 4981), (4982, 5077), (5078, 5163), (5164, 5240), (5241, 5371), (5372, 5415), (5416, 5621), (5622, 5738), (5739, 5954), (5955, 6166), (6167, 6254), (6255, 6391), (6392, 6553), (6554, 6613), (6614, 6770), (6771, 6956), (6957, 7268), (7269, 7444), (7445, 7604), (7605, 7691), (7692, 7735), (7736, 7856), (7857, 7909)]
[138, 223, 301]


In [216]:
# !pip install intervaltree

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
  Preparing metadata (setup.py) ... done
  Created wheel for intervaltree: filename=intervaltree-3.1.0-py2.py3-none-any.whl size=26094 sha256=0aa2c60800b9d3d47b0ab2ebcfdd1e4363bd26a9bdeff30d972f3d0abeaccf46
  Stored in directory: /Users/ameek/Library/Caches/pip/wheels/45/23/de/5789a92962483fd33cb06674792b9697c1b3766d7c7742830e
Successfully built intervaltree


In [357]:
# Code to identify sentence and highlight target - sending first instance of entity in a sentence to NewsSentiment. 
# Misses some sentences that are un-punctuated in the ULEZ research article.

from intervaltree import Interval, IntervalTree

print(clustered_entries[1]['Entity Name'])

START_HIGHLIGHT = '\033[0m'
END_HIGHLIGHT = '\033[94m'

GREEN = '\033[92m'
END_COLOR = '\033[0m'

cluster_positions = [(81, 91), (96, 99), (187, 199), (268, 280), (360, 366), (549, 579), (1006, 1013), (1230, 1255), (1360, 1367), (1479, 1503), (1735, 1745), (2099, 2108), (2503, 2510), (2669, 2670), (2688, 2689), (3393, 3400), (3741, 3746), (3771, 3773), (3796, 3798), (3866, 3869), (3936, 3938), (4136, 4146), (5326, 5342), (5352, 5355), (5915, 5924), (5999, 6010), (6154, 6157), (6342, 6349), (7103, 7110), (7216, 7225), (7468, 7479)]



bounds_tree = IntervalTree(Interval(start, end) for start, end in sentence_bounds)

processed_bounds = {}


for cluster_start, cluster_end in cluster_positions:
    overlap = bounds_tree.overlap(cluster_start, cluster_end)
    if overlap:
        first_unprocessed = None
        for interval in overlap:
            print(" ")
            sentence_start, sentence_end = interval.begin, interval.end
            bounds_key = (sentence_start, sentence_end)
            
            highlighted_text = (
                START_HIGHLIGHT + article_text[sentence_start:cluster_start] + END_HIGHLIGHT +
                article_text[cluster_start:cluster_end] +
                START_HIGHLIGHT + article_text[cluster_end:sentence_end] + END_HIGHLIGHT)
            print(START_HIGHLIGHT + f"Cluster ({cluster_start}, {cluster_end}) is within bounds ({sentence_start}, {sentence_end})")
            print(highlighted_text)
            
            
            if bounds_key not in processed_bounds:
                first_unprocessed = interval
                processed_bounds[bounds_key] = True
                process_subroutine(cluster_start, cluster_end, sentence_start, sentence_end)
                print(GREEN + "NewsSentiment Candidate as first bound appearance" + END_COLOR)
                break
        
       
        

Khan
 
Cluster (81, 91) is within bounds (0, 138)
'Think twice before imposing Ulez tax on hard-working Brits': Rishi Sunak blasts Sadiq Khan for his hated low emission zone expansion plan
{'class_id': 0, 'class_label': 'negative', 'class_prob': 0.990960419178009}
NewsSentiment Candidate as first bound appearance
 
Cluster (96, 99) is within bounds (0, 138)
'Think twice before imposing Ulez tax on hard-working Brits': Rishi Sunak blasts Sadiq Khan for his hated low emission zone expansion plan
 
Cluster (187, 199) is within bounds (139, 223)
- Keir Starmer told to 'get off the fence' over Sadiq Khan's Ulez by senior Tory MPs
{'class_id': 0, 'class_label': 'negative', 'class_prob': 0.9154871702194214}
NewsSentiment Candidate as first bound appearance
 
Cluster (268, 280) is within bounds (224, 301)
- Rishi Sunak led MPs asking Starmer to get London mayor to postpone the levy
{'class_id': 1, 'class_label': 'neutral', 'class_prob': 0.905670702457428}
NewsSentiment Candidate as first bound

In [347]:
# !pip install NewsSentiment

from NewsSentiment import TargetSentimentClassifier
tsc = TargetSentimentClassifier()

def process_subroutine(cluster_start, cluster_end, sentence_start, sentence_end):
    
    sentiment = tsc.infer_from_text(article_text[sentence_start:cluster_start] , article_text[cluster_start:cluster_end] , article_text[cluster_end:sentence_end])
    print(sentiment[0])

In [358]:
# Code to identify sentence and highlight target - sending every instance of entity in a sentence to NewsSentiment. 
# Keeps track of the number of sends to NewsSentiment so average can be taken.
from intervaltree import Interval, IntervalTree

START_HIGHLIGHT = '\033[0m'
END_HIGHLIGHT = '\033[94m'

GREEN = '\033[92m'
END_COLOR = '\033[0m'

cluster_positions = [(81, 91), (96, 99), (187, 199), (268, 280), (360, 366), (549, 579), (1006, 1013), (1230, 1255), (1360, 1367), (1479, 1503), (1735, 1745), (2099, 2108), (2503, 2510), (2669, 2670), (2688, 2689), (3393, 3400), (3741, 3746), (3771, 3773), (3796, 3798), (3866, 3869), (3936, 3938), (4136, 4146), (5326, 5342), (5352, 5355), (5915, 5924), (5999, 6010), (6154, 6157), (6342, 6349), (7103, 7110), (7216, 7225), (7468, 7479)]

bounds_tree = IntervalTree(Interval(start, end) for start, end in sentence_bounds)

processed_bounds_count = {}


    

for cluster_start, cluster_end in cluster_positions:
    overlap = bounds_tree.overlap(cluster_start, cluster_end)
    if overlap:
        for interval in overlap:
            print(" ")
            sentence_start, sentence_end = interval.begin, interval.end
            bounds_key = (sentence_start, sentence_end)
            
            if bounds_key not in processed_bounds_count:
                processed_bounds_count[bounds_key] = 0

            highlighted_text = (
            START_HIGHLIGHT + article_text[sentence_start:cluster_start] + END_HIGHLIGHT +
            article_text[cluster_start:cluster_end] +
            START_HIGHLIGHT + article_text[cluster_end:sentence_end] + END_HIGHLIGHT)
            print(START_HIGHLIGHT + f"Cluster ({cluster_start}, {cluster_end}) is within bounds ({sentence_start}, {sentence_end})")
            print(highlighted_text)
            
            process_subroutine(cluster_start, cluster_end, sentence_start, sentence_end)
            
            processed_bounds_count[bounds_key] += 1
            print(GREEN + f"NewsSentiment Candidate appearance {processed_bounds_count[bounds_key]}" + END_COLOR)

 
Cluster (81, 91) is within bounds (0, 138)
'Think twice before imposing Ulez tax on hard-working Brits': Rishi Sunak blasts Sadiq Khan for his hated low emission zone expansion plan
{'class_id': 0, 'class_label': 'negative', 'class_prob': 0.990960419178009}
NewsSentiment Candidate appearance 1
 
Cluster (96, 99) is within bounds (0, 138)
'Think twice before imposing Ulez tax on hard-working Brits': Rishi Sunak blasts Sadiq Khan for his hated low emission zone expansion plan
{'class_id': 0, 'class_label': 'negative', 'class_prob': 0.9810953736305237}
NewsSentiment Candidate appearance 2
 
Cluster (187, 199) is within bounds (139, 223)
- Keir Starmer told to 'get off the fence' over Sadiq Khan's Ulez by senior Tory MPs
{'class_id': 0, 'class_label': 'negative', 'class_prob': 0.9154871702194214}
NewsSentiment Candidate appearance 1
 
Cluster (268, 280) is within bounds (224, 301)
- Rishi Sunak led MPs asking Starmer to get London mayor to postpone the levy
{'class_id': 1, 'class_label':

In [163]:
collated_entities = {}

for entity_type, entities in entity_type_to_entities.items():
    for entity_info in entities:
        entity_text = entity_info[0]
        positions = entity_info[1]
        label = entity_info[2]

        if entity_text in collated_entities:
            if label in collated_entities[entity_text]:
                collated_entities[entity_text][label]['occurrences'] += len(positions)
                collated_entities[entity_text][label]['positions'].extend(positions)
            else:
                collated_entities[entity_text][label] = {'occurrences': len(positions), 'positions': positions}
        else:
            collated_entities[entity_text] = {label: {'occurrences': len(positions), 'positions': positions}}


entities_with_multiple_labels = {}

# for entity_text, label_occurrences in collated_entities.items():
#     output = [f"Entity Text: {entity_text}"]
#     for label, data in label_occurrences.items():
#         output.append(f"Label: {label}, Occurrences: {data['occurrences']}, Positions: {data['positions']}")
#     print(" | ".join(output))
    

# Print results where there are duplicate / in dispute NER labels:

# for entity_text, label_occurrences in collated_entities.items():
#     # Check if there are multiple labels associated with the entity text
#     if len(label_occurrences) > 1:
#         output = [f"Entity Text: {entity_text}"]
#         for label, data in label_occurrences.items():
#              output.append(f"Label: {label}, Occurrences: {data['occurrences']}, Positions: {data['positions']}")
#              # output.append(f"Label: {label}, Occurrences: {data['occurrences']}")
#         entities_with_multiple_labels[entity_text] = " | ".join(output)
#         print(" | ".join(output))
#         
# 
# 
# for entity_text, data in entities_with_multiple_labels.items():
#     print(f" {data}")

# Store results where there are duplicate / in dispute NER labels:
for entity_text, label_occurrences in collated_entities.items():
    if len(label_occurrences) > 1:
        entities_with_multiple_labels[entity_text] = []
        for label, data in label_occurrences.items():
            entity_info = {
                "Label": label,
                "Occurrences": data['occurrences'],
                "Positions": data['positions']
            }
            entities_with_multiple_labels[entity_text].append(entity_info)

# Print the stored alike entity text values that have multiple NER labels:
for entity_text, label_data in entities_with_multiple_labels.items():
    print("---------------------------")
    print(f"Entity Text: {entity_text}")
    for label_info in label_data:
        print(f"Label: {label_info['Label']}, Occurrences: {label_info['Occurrences']}, Positions: {label_info['Positions']}")


---------------------------
Entity Text: Ulez
Label: GPE, Occurrences: 4, Positions: [[2388, 2392], [3533, 3537], [4657, 4661], [6034, 6038]]
Label: ORG, Occurrences: 15, Positions: [[29, 33], [200, 204], [337, 341], [616, 620], [676, 680], [1573, 1577], [1639, 1643], [2627, 2631], [2967, 2971], [3342, 3346], [4182, 4186], [5356, 5360], [5650, 5654], [5925, 5929], [7350, 7354]]
Label: PERSON, Occurrences: 1, Positions: [[4200, 4204]]
---------------------------
Entity Text: Greater London
Label: GPE, Occurrences: 1, Positions: [[5671, 5685]]
Label: LOC, Occurrences: 1, Positions: [[2158, 2172]]
---------------------------
Entity Text: Tory
Label: NORP, Occurrences: 1, Positions: [[3207, 3211]]
Label: PERSON, Occurrences: 1, Positions: [[215, 219]]
---------------------------
Entity Text: Starmer
Label: ORG, Occurrences: 1, Positions: [[253, 260]]
Label: PERSON, Occurrences: 1, Positions: [[1688, 1695]]
---------------------------
Entity Text: Londoners
Label: ORG, Occurrences: 1, Posit

Appears using the majority case wouldn't always work. 

Let us park these inconclusive NER classifications here ^^^. After we have connected the NER to the coreference resolution output....

In [62]:
nlp = spacy.load("en_core_web_sm")

coreference_matrix = [
    'Sadiq Khan', 'his', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan",
    'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan',
    "Mr Khan's", 'Mr Khan', 'I', 'I', 'Mr Khan', 'Sadiq', 'He', 'he', 'his', 'he', 'your mayor',
    'the Labour mayor', 'his', "Mr Khan's", "the mayor's", 'his', 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"
]

split_matrix = [element.split() for element in coreference_matrix]

nouns = []
for words in split_matrix:
    for word in words:
        doc = nlp(word)
        for token in doc:
            if token.pos_ == "NOUN":
                nouns.append(token.text)

print(nouns)

['Sadiq', 'Sadiq', 'mayor', 'mayor', 'Sadiq', 'mayor', 'Sadiq', 'Sir', 'Keir', 'mayor', 'Sadiq', 'Sadiq', 'mayor', 'Labour', 'mayor', 'mayor', 'mayor', 'mayor']


Seems flawed approach as Keir has been brought out from the phrase "Sir Keir Starmer's mayor" even when we are referring to Sadiq Khan - additionally Khan was discarded too.

Discard a cluster c in a document d if |Mc| ≤ 0.2|Sd|, where |...| is the number of mentions of a cluster (Mc) and sentences in a document (Sd) (NEWS-MTSC approach)

In [63]:
for cluster in clustersText:
    print("-----------------------------")
    most_specific_mention = None
   
    text = " ".join(cluster)
    doc = nlp(text)

    for ent in doc.ents:
        if ent.label_ == "PERSON":
            most_specific_mention = most_specific_mention or ent.text

    if most_specific_mention:
        print(cluster)
        print("Most Specific PERSON Mention:", most_specific_mention)
    else:
        print(cluster)
        print("No specific PERSON name found in the cluster.")

-----------------------------
['Sadiq Khan', 'his', "Sadiq Khan's", 'London mayor', "Khan's", "the capital's mayor Sadiq Khan", 'Mr Khan', "London mayor Sadiq Khan's", 'Mr Khan', "Sir Keir Starmer's mayor", 'Sadiq Khan', "Mr Khan's", 'Mr Khan', 'I', 'I', 'Mr Khan', 'Sadiq', 'He', 'he', 'his', 'he', 'your mayor', 'the Labour mayor', 'his', "Mr Khan's", "the mayor's", 'his', 'Mr Khan', 'Mr Khan', 'the mayor', "the mayor's"]
Most Specific PERSON Mention: Khan
-----------------------------
['Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak', 'Rishi Sunak (pictured)']
Most Specific PERSON Mention: Rishi
-----------------------------
['Keir Starmer', 'Starmer', 'Keir Starmer', 'him', 'he', 'Sir Keir', 'He', 'the Labour leader', 'Keir Starmer (left)', 'him', 'the Labour leader', "Sir Keir Starmer's", 'Starmer', 'Sir Keir', 'the Labour leader', 'Keir', 'he', 'Sir Keir', 'Sir Keir']
Most Specific PERSON Mention: Keir Starmer Starmer Keir Starmer
-----------------------------
["Sadiq Khan's Ulez", 'ULE

In [6]:
# TODO: Resolve conflict code above to identify prevalence of NER unsure.
# TODO: Match co-reference resolution outcomes with NER identification to increase positions code.
# TODO: WIKIPEDIA Resolution? (Nice to have? for summaries?)
# TODO: Extract sentences for all identified mentions
# TODO: Apply NewsSentiment with correct start/end positions and formatting resolution